src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,330 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.FeatureManagement.FeatureFilters
+{
+    static class RecurrenceEvaluator
+    {
+        const int DaysPerWeek = 7;
+
+        /// <summary>
+        /// Checks if a provided timestamp is within any recurring time window specified by the Recurrence section in the time window filter settings.
+        /// If the time window filter has an invalid recurrence setting, an exception will be thrown.
+        /// <param name="time">A timestamp.</param>
+        /// <param name="settings">The settings of time window filter.</param>
+        /// <returns>True if the timestamp is within any recurring time window, false otherwise.</returns>
+        /// </summary>
+        public static bool MatchRecurrence(DateTimeOffset time, TimeWindowFilterSettings settings)
+        {
+            if (time < settings.Start.Value)
+            {
+                return false;
+            }
+
+            if (TryFindPreviousOccurrence(time, settings, out DateTimeOffset previousOccurrence, out int _))
+            {
+                return time < previousOccurrence + (settings.End.Value - settings.Start.Value);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Try to find the closest previous recurrence occurrence (if any) before the provided timestamp and the next occurrence.
+        /// <param name="time">A timestamp.</param>
+        /// <param name="settings">The settings of time window filter.</param>
+        /// <param name="prevOccurrence">The closest previous occurrence. If there is no previous occurrence, it will be set to <see cref="DateTimeOffset.MinValue"/>.</param>
+        /// <param name="nextOccurrence">The next occurrence.</param>
+        /// <returns>True if the closest previous occurrence is within the recurrence range or the time is before the first occurrence, false otherwise.</returns>
+        /// </summary>
+        public static bool TryFindPrevAndNextOccurrences(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset prevOccurrence, out DateTimeOffset nextOccurrence)
```

Comment:
This method is really hard to understand. Given the current name, one would expect it to only be true if prevOccurrence and nextOccurrence are both found.

Reply:
Or

```
GetSurroundingOccurrences(
    DateTimeOffset time,
    out DateTimeOffset? previous,
    out DateTimeOffset? next
)
```

Reply:
How about rename it to `TryCalculateSurroundingOccurrences(
    DateTimeOffset time,
    TimeWindowFilterSetting settings,
    out DateTimeOffset previous,
    out DateTimeOffset next)
    
This method will be true if either prev or next occurrence is found.

Reply:
Updated.

``` C#
/// <summary>
/// Calculate the closest previous recurrence occurrence (if any) before the provided timestamp and the next occurrence (if any) after the provided timestamp.
/// <param name="time">A timestamp.</param>
/// <param name="settings">The settings of time window filter.</param>
/// <param name="prevOccurrence">The closest previous occurrence. Note that prev occurrence can be null even if the time is past the start date, because the recurrence range may have surpassed its end.</param>
/// <param name="nextOccurrence">The next occurrence. Note that next occurrence can be null even if the prev occurrence is not null, because the recurrence range may have reached its end.</param>
/// </summary>
public static void CalculateSurroundingOccurrences(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset? prevOccurrence, out DateTimeOffset? nextOccurrence)
{
    prevOccurrence = null;

    nextOccurrence = null;

    if (time < settings.Start.Value)
    {
        nextOccurrence = settings.Start.Value;

        return;
    }

    if (TryFindPreviousOccurrence(time, settings, out DateTimeOffset prev, out int numberOfOccurrences))
    {
        prevOccurrence = prev;

        RecurrencePattern pattern = settings.Recurrence.Pattern;

        if (pattern.Type == RecurrencePatternType.Daily)
        {
            nextOccurrence = prev.AddDays(pattern.Interval);
        }

        if (pattern.Type == RecurrencePatternType.Weekly)
        {
            nextOccurrence = CalculateWeeklyNextOccurrence(prev, settings);
        }

        RecurrenceRange range = settings.Recurrence.Range;

        if (range.Type == RecurrenceRangeType.EndDate)
        {
            if (nextOccurrence > range.EndDate)
            {
                nextOccurrence = null;
            }
        }

        if (range.Type == RecurrenceRangeType.Numbered)
        {
            if (numberOfOccurrences >= range.NumberOfOccurrences)
            {
                nextOccurrence = null;
            }
        }
    }
}
```